src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,257 @@
+/*-------------------------------------------------------------------------
+ *
+ * intermediate_result_pruning.c
+ *   Functions for pruning intermediate result broadcasting.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "distributed/citus_custom_scan.h"
+#include "distributed/intermediate_result_pruning.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/worker_manager.h"
+#include "utils/builtins.h"
+
+static List * AppendAllAccessedWorkerNodes(List *workerNodeList,
+										   DistributedPlan *distributedPlan,
+										   int workerNodeCount);
+static IntermediateResultsHashEntry * SearchIntermediateResult(HTAB
+															   *intermediateResultsHash,
+															   char *resultId);
+
+
+/*
+ * FindSubPlansUsedInPlan adds all the subplans used by the plan into
+ * plan->usedSubPlanNodeList field. Simply traversing the range table entries in the plan.
+ */
+void
+FindSubPlansUsedInPlan(DistributedPlan *plan)
+{
+	Query *jobQuery = NULL;
+	List *rangeTableList = NIL;
+	ListCell *rangeTableCell = NULL;
+
+	jobQuery = plan->workerJob == NULL ? plan->insertSelectSubquery :
```

Comment:
Don't forget to use `originalQuery` instead of `jobQuery`

Reply:
bump

Reply:
We lose some of the information when we use originalQuery, and end up not sending the intermediate results to some of the workers.

You can see one failed attempt here: https://github.com/citusdata/citus/tree/hanefi-failing-originalquery